src/solve.jl

```diff
@@ -17,71 +17,493 @@ function DiffEqBase.__solve(prob::BVProblem, alg::Shooting; kwargs...)
+                    new_Y = zeros(Nsub_star + 1, len)
+                    for i in 0:Nsub_star
+                        z, z_prime = interp_eval(mesh, Y, alg, mesh_new[i + 1], dt, len, s, s_star, k_discrete, k_interp)
+                        new_Y[i + 1, :] = z
+                    end
+                    mesh = copy(mesh_new)
+                    n = copy(Nsub_star)
+                    Y = copy(new_Y)
+                end
+            end
+
+            if (info == 0) && (defect_norm < tol)
+                println("Succesful computation, the user defined tolerance has been satisfied")
+            end
+        else
+            println("Cannot obtain a solution for the current mesh")
+            if 2 * n > MxNsub
+                println("New mesh would be too large")
+                info = -1
+            else
+                mesh_new = half_mesh(mesh, n)
+                Nsub_star = 2 * n
+                mesh = copy(mesh_new)
+                n = copy(Nsub_star)
+                println("New mesh will be of size ", n) # Next computation would be based on length n mesh
+                info = 0 # Force a restart
+                defect_norm = 2 * tol
+            end
+        end
     end
-    function loss!(resid, u0, p)
-        nest_vector!(S.y, u0)
-        @set! S.p = p
-        Φ!(S, tableau, cache)
-        if isa(prob.problem_type, TwoPointBVProblem)
-            eval_bc_residual!(S)
+
+    retcode = ReturnCode.Success
+    if info == 0
+        retcode = ReturnCode.Success
+    elseif info == -4
+        retcode = ReturnCode.Terminated
+    else
+        retcode = ReturnCode.Failure
+    end
+   return DiffEqBase.build_solution(prob, alg, x, S.y; opt.retcode)
+end
+
+"""
+    interp_eval(mesh, Y, t, k_discrete, k_interp)
+
+After we construct an interpolant, we use interp_eval to evaluate it.
+"""
+function interp_eval(mesh, Y, alg, t, dt, len, s, s_star, k_discrete, k_interp)
+    # EXPORTS: z, z_prime
+    i = interval(mesh, t)
+    hi = mesh[i + 1] - mesh[i]
+    tau = (t - mesh[i]) / hi
+    weights, weights_prime = interp_weights(tau, alg)
+    z, z_prime = sum_stages(weights, weights_prime, k_discrete, k_interp, len, dt, Y, s,
+                            s_star)
+    return z, z_prime
+end
+
+function interval(mesh, t)
+    ind = findfirst(x -> x > t, mesh)
+    i::Int64 = copy(ind)
+    return i
+end
+
+"""
+    mesh_selector(mesh_current, defect, tol, n, len, alg)
+
+Generate new mesh based on the defect.
+"""
+function mesh_selector(mesh_current::Vector, defect, tol, n::Int64, len::Int64,
+                       alg::Union{GeneralMIRK, MIRK})
+    #exports: mesh_new, Nsub_star, info
+
+    #TODO: Need users to manually specify, here, we set it as 3000 by default.
+    MxNsub = 3000
+
+    safety_factor = 1.3
+    rho = 1.0 # Set rho=1 means mesh distribution will take place everytime.
+    upper_new_mesh = 4.0
+    lower_new_mesh = 0.5
+    r1 = 0.0
+    r2 = 0.0
+    Nsub_star = 0
+    info = 0
```

Comment:
are these pulled from the paper?

Reply:
The BVP book does have the specification of r1, r2 and r3

![image](https://github.com/SciML/BoundaryValueDiffEq.jl/assets/52615090/0faade0d-8147-4505-b1dd-1720397a4420)


But both the book and the paper don't elaborate on the setting of the other parameters, so I set these parameters according to the FORTRAN routine.